sphinx/ext/autodoc.py

```diff
@@ -508,6 +513,36 @@ def parse_name(self):
                         (self.objpath and '.' + '.'.join(self.objpath) or '')
         return True
 
+    def monkey_patch_import_function(self):
+        """Replace the builtin ``import`` function.
+
+        When a module cannot be imported, and its name begins by something it
+        ``autodoc_mock_imports``, return a dummy one which makes the
+        auto-documented modules "importable".
+        """
+        def _import(name, globals=None, locals=None, fromlist=(), level=None):
+            try:
+                if PY2:
+                    if globals is None:
+                        globals = {}
+                    if locals is None:
+                        locals = {}
+                    if level is None:
+                        level = -1
+                elif level is None:
+                    level = 0
+                return REALIMPORT(name, globals, locals, fromlist, level)
+            except ImportError:
+                if any(name.startswith(m)
+                       for m in self.env.config.autodoc_mock_imports):
+                    self.env.app.debug('[autodoc] adding a mock module %s!'
+                                       % name)
+                    return _MockModule(name)
+                else:
+                    raise
+
+        builtins.__import__ = _import
```

Comment:
Please use import hook mechanism: https://docs.python.org/3/reference/import.html

Before python-2.3, we have to replace `__import__` method to hook importing mechanism. However [PEP-0302](https://www.python.org/dev/peps/pep-0302/) provides a way to hook importing.
[PEP-0451](https://www.python.org/dev/peps/pep-0451/) is latest specification of importing, however (I think) PEP-0451 is not applied to Python-2.7.

My conclusion is, we can implement the purpose of this PR with using PEP-0302 spec for both of Py2 and Py3.